File: src/sdk.ts
import type {
  ContextLaneRun,
  ContextLaneRunMeta,
  ContextLaneExtraction,
  ContextLaneMemoryRecord,
} from './core/schema.js'
import type { SearchResult } from './core/search.js'

export interface ClientOptions {
  baseUrl: string
  apiKey?: string
}

export interface IngestRequest {
  input: string
  type?: 'file' | 'folder' | 'url' | 'github' | 'text'
  syncToMemoryLane?: boolean
}

export interface RecallRequest {
  query: string
  limit?: number
}

export class ContextLaneClient {
  private baseUrl: string
  private headers: Record<string, string>

  constructor(opts: ClientOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/+$/, '')
    this.headers = { 'Content-Type': 'application/json' }
    if (opts.apiKey) {
      this.headers['Authorization'] = `Bearer ${opts.apiKey}`
    }
  }

  async health(): Promise<{ status: string }> {
    const res = await fetch(`${this.baseUrl}/health`)
    return res.json()
  }

  async capabilities(): Promise<Record<string, unknown>> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/capabilities`)
    return res.json()
  }

  async ingest(req: IngestRequest): Promise<ContextLaneRun> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/ingest`, {
      method: 'POST', headers: this.headers, body: JSON.stringify(req),
    })
    if (!res.ok) throw new Error(await res.text())
    return res.json()
  }

  async ingestUrl(url: string, syncToMemoryLane?: boolean): Promise<ContextLaneRun> {
    return this.ingest({ input: url, type: 'url', syncToMemoryLane })
  }

  async ingestGithub(repoUrl: string, syncToMemoryLane?: boolean): Promise<ContextLaneRun> {
    return this.ingest({ input: repoUrl, type: 'github', syncToMemoryLane })
  }

  async listRuns(): Promise<ContextLaneRunMeta[]> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/runs`, { headers: this.headers })
    return res.json()
  }

  async getRun(id: string): Promise<ContextLaneRun> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/runs/${id}`, { headers: this.headers })
    return res.json()
  }

  async getReport(id: string): Promise<string> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/runs/${id}/report`, { headers: this.headers })
    return res.text()
  }

  async getMemoryRecords(id: string): Promise<ContextLaneMemoryRecord[]> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/runs/${id}/memory-records`, { headers: this.headers })
    return res.json()
  }

  async search(query: string, limit = 5): Promise<SearchResult[]> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/search`, {
      method: 'POST', headers: this.headers, body: JSON.stringify({ query, limit }),
    })
    return res.json()
  }

  async recall(query: string, limit = 5): Promise<SearchResult[]> {
    return this.search(query, limit)
  }

  async syncMemoryLane(runId: string): Promise<{ saved: number; failed: number; method: string }> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/sync/memorylane`, {
      method: 'POST', headers: this.headers, body: JSON.stringify({ runId }),
    })
    return res.json()
  }

  async exportRun(runId: string, outPath: string): Promise<void> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/export`, {
      method: 'POST', headers: this.headers, body: JSON.stringify({ runId, outPath }),
    })
    if (!res.ok) throw new Error(await res.text())
  }

  async importRun(dataPath: string): Promise<ContextLaneRun> {
    const res = await fetch(`${this.baseUrl}/v1/contextlane/import`, {
      method: 'POST', headers: this.headers, body: JSON.stringify({ path: dataPath }),
    })
    return res.json()
  }
}

export function createContextLaneClient(opts: ClientOptions): ContextLaneClient {
  return new ContextLaneClient(opts)
}

File: python/contextlane/sdk.py
import json
import urllib.request
import urllib.error


class ContextLaneError(Exception):
    pass


class ContextLaneClient:
    def __init__(self, base_url: str, api_key: str | None = None):
        self.base_url = base_url.rstrip("/")
        self.headers = {"Content-Type": "application/json"}
        if api_key:
            self.headers["Authorization"] = f"Bearer {api_key}"

    def _request(self, method: str, path: str, body: dict | None = None):
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode("utf-8") if body else None
        req = urllib.request.Request(url, data=data, headers=self.headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                ct = resp.headers.get("Content-Type", "")
                text = resp.read().decode("utf-8")
                if "application/json" in ct:
                    return json.loads(text)
                return text
        except urllib.error.HTTPError as e:
            raise ContextLaneError(f"HTTP {e.code}: {e.read().decode()}")
        except urllib.error.URLError as e:
            raise ContextLaneError(f"URL error: {e.reason}")

    def health(self) -> dict:
        return self._request("GET", "/health")

    def capabilities(self) -> dict:
        return self._request("GET", "/v1/contextlane/capabilities")

    def ingest(self, input_str: str, type: str | None = None,
               sync_to_memory_lane: bool = False) -> dict:
        return self._request("POST", "/v1/contextlane/ingest", {
            "input": input_str, "type": type,
            "syncToMemoryLane": sync_to_memory_lane,
        })

    def ingest_url(self, url: str, sync_to_memory_lane: bool = False) -> dict:
        return self.ingest(url, type="url", sync_to_memory_lane=sync_to_memory_lane)

    def ingest_github(self, repo_url: str, sync_to_memory_lane: bool = False) -> dict:
        return self.ingest(repo_url, type="github", sync_to_memory_lane=sync_to_memory_lane)

    def list_runs(self) -> list:
        return self._request("GET", "/v1/contextlane/runs")

    def get_run(self, run_id: str) -> dict:
        return self._request("GET", f"/v1/contextlane/runs/{run_id}")

    def get_report(self, run_id: str) -> str:
        return self._request("GET", f"/v1/contextlane/runs/{run_id}/report")

    def get_memory_records(self, run_id: str) -> list:
        return self._request("GET", f"/v1/contextlane/runs/{run_id}/memory-records")

    def search(self, query: str, limit: int = 5) -> list:
        return self._request("POST", "/v1/contextlane/search", {
            "query": query, "limit": limit,
        })

    def recall(self, query: str, limit: int = 5) -> list:
        return self.search(query, limit)

    def sync_memory_lane(self, run_id: str) -> dict:
        return self._request("POST", "/v1/contextlane/sync/memorylane", {
            "runId": run_id,
        })

    def export_run(self, run_id: str, out_path: str) -> None:
        self._request("POST", "/v1/contextlane/export", {
            "runId": run_id, "outPath": out_path,
        })

    def import_run(self, data_path: str) -> dict:
        return self._request("POST", "/v1/contextlane/import", {
            "path": data_path,
        })


def create_context_lane_client(base_url: str, api_key: str | None = None) -> ContextLaneClient:
    return ContextLaneClient(base_url, api_key)
